cpu/em_quad_level_frequency_transitions.py

```diff
@@ -0,0 +1,157 @@
+
+    def test(self):
+        """
+        1) Change governor to userspace on all CPUs.
+        2) For each core in the quad, pick a random frequency
+        and set cpu with that frequency.
+        3) Validate that the cpuinfo_cur_freq on any core in
+        the code is set to max(set of frequencies)
+        4) Run the perf command on each cpu to validate frequencies
+        independently.
+        """
+        threshold = 300000
+        self.quad_dict = {}
+        self.max_freq_dict = {}
+        self.quad_to_cpu_mapping()
+        output = process.run("cpupower frequency-set -g userspace", shell=True)
+        cur_governor = self.cpu_freq_path('scaling_governor')
+        if 'userspace' == cur_governor and output.exit_status == 0:
+            self.log.info("%s governor set successfully" % cur_governor)
+        else:
+            self.cancel("Unable to set the userspace governor")
+        for chip in self.quad_dict:
+            for quad in self.quad_dict[chip]:
+                for self.cpu_num in self.quad_dict[chip][quad]:
+                    process.run("cpupower -c %s frequency-set -f %s"
+                                % (self.cpu_num, self.get_random_freq()))
+                    time.sleep(1)
+                    freq_set = self.cpu_freq_path('cpuinfo_cur_freq')
+                    if self.max_freq < freq_set:
+                        self.max_freq = freq_set
+                if chip not in self.max_freq_dict.keys():
+                    self.max_freq_dict[chip] = {}
+                if quad not in self.max_freq_dict[chip].keys():
+                    self.max_freq_dict[chip][quad] = self.max_freq
+                    self.log.info("Maximum frequency set:%s quad:"
+                                  "%s" % (self.max_freq, quad))
+                self.max_freq = 0
+        for chip in self.quad_dict:
+            for quad in self.quad_dict[chip]:
+                for cpu in self.quad_dict[chip][quad]:
+                    freq_get = self.perf_cmd(cpu)
+                    freq_max = self.max_freq_dict[chip][quad]
+                    diff = float(freq_max) - float(freq_get)
+                    if diff > threshold:
+                        self.cancel("Quad level max frequency %s is not set on"
+                                    "this cpu %s"
+                                    % (self.max_freq_dict[chip][quad], cpu))
+            self.log.info("Quad level max frequency %s is set on this quad"
+                          "%s" % (self.max_freq_dict[chip][quad], quad))
+
+    def quad_to_cpu_mapping(self):
+        """
+        Get the total quad and cpus list belonging to each quad.
+        """
+        self.nums = [x for x in range(0, self.cpus)]
+        for cpu in self.nums:
+            phy_id = process.system_output('cat /sys/devices/system/'
+                                           'cpu/cpu%s/physical_id' % cpu)
+            quad_id = int(phy_id) >> 4 & 0x7
+            chip_id = int(phy_id) >> 8 & 0x7F
```

Comment:
lost here :( , can you please help me to understand  why these operation require ?

Reply:
On P9, we have 6 quads and each quad has cores . but to get which cores belong to which quad, we need to calculate using the above formula. When change is made to frequency , it is reflected at quad level hence this calculation is needed